plugins/profiles.koplugin/main.lua

```diff
@@ -0,0 +1,128 @@
+end
+
+function Profiles:onFlushSettings()
+    if self.profiles then
+        self.profiles:flush()
+    end
+end
+
+function Profiles:addToMainMenu(menu_items)
+    logger.dbg("Profiles:addToMainMenu")
+    menu_items.profiles = {
+        text = _("Profiles"),
+        sub_item_table_func = function()
+            return self:getSubMenuItems()
+        end,
+    }
+end
+
+function Profiles:getSubMenuItems()
+    self:loadProfiles()
+    local sub_item_table = {
+        {
+            text = _("New"),
+            keep_menu_open = true,
+            callback = function(touchmenu_instance)
+                local name_input
+                name_input = InputDialog:new{
+                    title =  _("Enter profile name"),
+                    input = "",
+                    buttons = {{
+                        {
+                            text = _("Cancel"),
+                            callback = function()
+                                UIManager:close(name_input)
+                            end,
+                        },
+                        {
+                            text = _("Save"),
+                            callback = function()
+                                local name = name_input:getInputText()
+                                UIManager:close(name_input)
+                                self:newProfile(name)
+                                touchmenu_instance.item_table = self:getSubMenuItems()
+                                touchmenu_instance.page = 1
+                                touchmenu_instance:updateItems()
+                            end,
+                        },
+                    }},
+                }
+                UIManager:show(name_input)
+                name_input:onShowKeyboard()
+            end,
+            separator = true,
+        }
+    }
+    for k,v in pairs(self.data) do
+        local sub_items = {
+            {
+                text = _("Delete profile"),
+                keep_menu_open = false,
+                separator = true,
+                callback = function()
+                    UIManager:show(ConfirmBox:new{
+                        text = _("Do you want to delete this profile?"),
+                        ok_text = _("Yes"),
+                        cancel_text = _("No"),
+                        ok_callback = function()
+                            self:deleteProfile(k)
+                        end,
+                    })
+                end,
+            }
+        }
+        Dispatcher.addSubMenu(self, sub_items, "data", k)
+        table.insert(sub_item_table, {
+            text = k,
+            hold_keep_menu_open = false,
+            sub_item_table = sub_items,
+            hold_callback = function()
+                logger.dbg("Profile callback")
+                Dispatcher.execute(self, v)
+            end,
+        })
+    end
+    return sub_item_table
+end
+
+function Profiles:newProfile(name)
+    if self.data[name] == nil then
```

Comment:
More of a minor detail but it seems like where this is used it should perhaps warn that it already exists or something?

Reply:
If the profile exists, it will do nothing, leaving the old profile as is.

Reply:
Right, it's a minor detail because it's not dangerous, but the effect is the same as cancel.

Reply:
It is in the callback, so it would be hard to  bring them back to the input dialog (the best idea), and I think a warning will just scare a user unnecessarily.

Reply:
I'm not sure I follow, just don't close the dialog? Along these lines:

```lua
                                local name = name_input:getInputText()
                                if not self:newProfile(name) then
                                    -- InfoMessage or something
                                    return
                                end
                                UIManager:close(name_input)
```

Reply:
wasnt thinking :)